### ISSUE TITLE
Help with pool channels

### ISSUE THREAD

**Issue by n1cod3mus** [samphonic/MoppyClassic#65]
I have Moppy Advanced setup and i have been playing with it a far bit and throwing various midi files I downloaded off the net at it. Now i know the tone range is limited so i may not get all the notes.

I have only 4 drives set up right now (I'm going to add more when i come across them) which means anything in the other channels on a midi will be missed. I did notice that there is Drive-Pooling, now i assume that forces the program to take say 8 channels and play the notes across only 4 channels.

only thing is, I have no idea how to use those settings as there is 2 drop downs for pool channels, and 2 drop downs for Into Channels. It would have made more sense to have 1 drop down box for each surly. or am I missing the purpose of the second drop down.

in short, i need help in how to use that area of the interface and what it actually does, so far i have just assumed what it does as there is nothing on the readme about using the interface and there is nothing I can find by googling or seaching this page on how to use this part. oh an what is the Strategy for? i assume its how it allocates the notes across the drives/channels?

thanks in advance for your help.

**Comment by MultiDaxio** [comment]
Hmmm... Drive pooling is quite useful indeed. You might try changing these
options above (straight forward, round robin etc.) Lemme know if it does
the trick.
-MultiDaxio

**Comment by n1cod3mus** [comment]
ok thanks, but how do I use all the drop downs properly? I'm just guessing at the moment I have no idea what I'm doing

**Comment by n1cod3mus** [comment]
I have changed the various options such as round-robin but i dont really notice much difference

**Comment by n1cod3mus** [comment]
ooooooooohhhh i am a numpty, i get it, because you can only have 8 drives on a uno, the first drop down is for the first UNO and the 2nd drop down is for the 2nd UNO?? am i right?

**Comment by MultiDaxio** [comment]
Yup :)

**Comment by n1cod3mus** [comment]
ok I have no got 2 UNOs with 8 drives on each, but I'm not actually sure of how to use the pooling properly.

say for example I have a MIDI that only has 8 channels, how would i get stack B to do the same as stack A?

and how else can the pooling be used?

**Comment by MultiDaxio** [comment]
Honestly, I have not used pooling that much to know it. Try asking Sammy1am
or just google it :)

**Comment by n1cod3mus** [comment]
I was hoping Sammy might read this and reply, and honestly i tried to google moppy pooling before and it wasnt helpful

**Comment by MultiDaxio** [comment]
Be patient, my friend. Sammy will read it eventually. Sooner or later :)

**Comment by n1cod3mus** [comment]
its just bugging me as all i can do is play a midi and if it happens to use channels 9 - 16 then they play. But i have some that are 8 channel which would be awesome if i could just mirror the stacks in some way.

Its like about 6 of the drives rarely get use.

**Comment by MultiDaxio** [comment]
If you want to find use for your unused drives, you can use program like
SpeedyMIDI or other midi creators and add more tracks (you can also copy
notes from drives that play drums-line or vocal-line to make the sound
better)

**Comment by n1cod3mus** [comment]
ok i will look in to it thanks. 

is there any software out there that would allow you to merge channels? or can i do that in this speedyMIDI?

**Comment by MultiDaxio** [comment]
SpeedyMIDI is an universal MIDI editor, I am using myself to give my MIDIs
more that oomph! that power!

**Comment by samphonic** [comment]
Haha, I've been keeping tabs on the conversation, but it kinda sounded like you guys had things figured out there earlier in the week.

To be entirely honest, I'd sort of forgotten how the UI was laid out, so I had to go check on it again, but I think I remember now.  =P  Pooling was originally designed to allow a group of drives to play a MIDI with simultaneous notes in a single track-- mostly so that if you hooked a MIDI keyboard up to the drives, you could play as many notes simultaneously as you had drives.

For these examples, I'm going to assume a setup that has just 4 drives (just so I don't have to type as many numbers), but it should work with any number.

The "round-robin" pooling strategy assigns each incoming note to the _next_ drive, and wraps around after the last one: 1-2-3-4-1-2-3-4-1-2-&c....  The "stacking" strategy will assign notes to the _first_ available drive until all the drives are playing a note.  So for example, a simple scale would just play all the notes on drive 1, but a three note chord would use drives 1-3.  If 5 simultaneous notes show up, only the first 4 would play and the 5th would be ignored (at least that's what it SHOULD do... reviewing the code, it looks like "stacking" might be broken).

You've got a few options in your situation:
- If you're really set on using pooling, you can set the "into channels" to 1-16, and the "pool channels" to the channels in your MIDI file that actually have notes.  If you choose "round-robin", all 16 drives _should_ be used to play notes, but unless any of the incoming channels have simultaneous notes, all 16 will never end up playing at the same time.  Also, you lose the effect of having a specific drive play a specific part of the song-- so the melody will be spread throughout your drives.  Kinda cool, but not always what you want.
- You'll get the best control over the whole situation by just editing your MIDI files to include 16 channels and not using pooling at all.  If you simply want everything doubled, it shouldn't be too hard to use a MIDI editor to just duplicate the original tracks and assign them to all the channels.
- If you're feeling adventurous (and don't want to have to deal with editing the MIDI files every single time) you can git the code and add your own pooling strategy; basically, you just need to create a `useAllDrivesMap(MidiMessage message)` method in `DrivePooler.java` that does what you want it to.  You could have it automatically duplicate incoming tracks to fill up all the available output tracks.

Hopefully this helps get you on your way.  I would love to just implement a new pooling strategy myself for you, but I don't know when I'd have the time to get the drives out and test it.  If you DO decide to implement your own though, I'm more than happy to help with any questions about the code.

Good luck!

**Comment by MultiDaxio** [comment]
Ahh Sammy... You are a flippin ninja. Just when I have dried out of ideas,
you came in. TEACH MEH MASTAH! Ahem... Sorry. So basicaly you are telling
me that drive-pooling is sorta broken in Moppy app? If so, why you just do
not get rid of it in the interface? Lack of time? I am not ordering you,of
course not, but useless (partialy) functions I think should be terminated.
-MultiDaxio

2014-05-02 19:58 GMT+02:00 SammyIAm notifications@github.com:

> Haha, I've been keeping tabs on the conversation, but it kinda sounded
> like you guys had things figured out there earlier in the week.
> 
> To be entirely honest, I'd sort of forgotten how the UI was laid out, so I
> had to go check on it again, but I think I remember now. =P Pooling was
> originally designed to allow a group of drives to play a MIDI with
> simultaneous notes in a single track-- mostly so that if you hooked a MIDI
> keyboard up to the drives, you could play as many notes simultaneously as
> you had drives.
> 
> For these examples, I'm going to assume a setup that has just 4 drives
> (just so I don't have to type as many numbers), but it should work with any
> number.
> 
> The "round-robin" pooling strategy assigns each incoming note to the
> _next_ drive, and wraps around after the last one:
> 1-2-3-4-1-2-3-4-1-2-&c.... The "stacking" strategy will assign notes to the
> _first_ available drive until all the drives are playing a note. So for
> example, a simple scale would just play all the notes on drive 1, but a
> three note chord would use drives 1-3. If 5 simultaneous notes show up,
> only the first 4 would play and the 5th would be ignored (at least that's
> what it SHOULD do... reviewing the code, it looks like "stacking" might be
> broken).
> 
> You've got a few options in your situation:
> - If you're really set on using pooling, you can set the "into
>   channels" to 1-16, and the "pool channels" to the channels in your MIDI
>   file that actually have notes. If you choose "round-robin", all 16 drives
>   _should_ be used to play notes, but unless any of the incoming
>   channels have simultaneous notes, all 16 will never end up playing at the
>   same time. Also, you lose the effect of having a specific drive play a
>   specific part of the song-- so the melody will be spread throughout your
>   drives. Kinda cool, but not always what you want.
> - You'll get the best control over the whole situation by just editing
>   your MIDI files to include 16 channels and not using pooling at all. If you
>   simply want everything doubled, it shouldn't be too hard to use a MIDI
>   editor to just duplicate the original tracks and assign them to all the
>   channels.
> - If you're feeling adventurous (and don't want to have to deal with
>   editing the MIDI files every single time) you can git the code and add your
>   own pooling strategy; basically, you just need to create a useAllDrivesMap(MidiMessage
>   message) method in DrivePooler.java that does what you want it to. You
>   could have it automatically duplicate incoming tracks to fill up all the
>   available output tracks.
> 
> Hopefully this helps get you on your way. I would love to just implement a
> new pooling strategy myself for you, but I don't know when I'd have the
> time to get the drives out and test it. If you DO decide to implement your
> own though, I'm more than happy to help with any questions about the code.
> 
> Good luck!
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/SammyIAm/Moppy/issues/65#issuecomment-42059640
> .

## 

Adam

**Comment by n1cod3mus** [comment]
ok i think that has explained pooling, so the drop downs are for the start channel and the end channel. so if i wanted to pool only channels 1 - 4 over 16 channels, i would set Pool Channels drop downs to 1 and 4, and set the into channels to 1 and 16? then all the additional notes that over lap would be played on a drive that is not currently in use.

**Comment by MultiDaxio** [comment]
I think that is how it works :)
-MultiDaxio

2014-05-02 20:13 GMT+02:00 n1cod3mus notifications@github.com:

> ok i think that has explained pooling, so the drop downs are for the start
> channel and the end channel. so if i wanted to pool only channels 1 - 4
> over 16 channels, i would set Pool Channels drop downs to 1 and 4, and set
> the into channels to 1 and 16? then all the additional notes that over lap
> would be played on a drive that is not currently in use.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/SammyIAm/Moppy/issues/65#issuecomment-42061143
> .

## 

Adam

**Comment by n1cod3mus** [comment]
and the only way to mirror the stack would be do it in in a midi editor. 

A thought does occur, if you know that a channel is not in use, you could set that channel to mirror another channel which would have data, so you could set channel 14 to mirror channel 1, then anytime channel 1 plays anything so does channel 14

**Comment by MultiDaxio** [comment]
Basicly: floppies game of copy-cat :D Mirrored floppies are copying
themselves.

2014-05-02 20:15 GMT+02:00 n1cod3mus notifications@github.com:

> and the only way to mirror the stack would be do it in in a midi editor.
> 
> A thought does occur, if you know that a channel is not in use, you could
> set that channel to mirror another channel which would have data, so you
> could set channel 14 to mirror channel 1, then anytime channel 1 plays
> anything so does channel 14
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/SammyIAm/Moppy/issues/65#issuecomment-42061400
> .

## 

Adam

**Comment by samphonic** [comment]
Drive pooling's not broken, just that one pooling strategy appears to have a bug.  Otherwise it works just as intended.  =)  If I had time to work on it, I'd fix it rather than just removing it.

The mirroring that you're talking about, @n1cod3mus , is the sort of thing that could be implemented as a pooling strategy (see the third option I listed).

**Comment by n1cod3mus** [comment]
yeah basically, then you could modify the interface to have almost unlimited drives, and build a floppy organ that was like 50, 100, 500 floppy drives, but they would be mirrored

how epic would that be, you could do a world record, imagine the sound of 500 floppys only playing 8 channels

**Comment by MultiDaxio** [comment]
Wait a bit... One Arduino hands 8 floppies... 500/8 = 62.5 = around 63
1 Arduino = 35$
63 Arduinos = 2205$
500 Floppies = 15$ \* 500 = 7500$
Sum that... God... SHUT UP AND TAKE MY MONEY :)

2014-05-02 20:19 GMT+02:00 n1cod3mus notifications@github.com:

> yeah basically, then you could modify the interface to have almost
> unlimited drives, and build a floppy organ that was like 50, 100, 500
> floppy drives, but they would be mirrored
> 
> how epic would that be, you could do a world record, imagine the sound of
> 500 floppys only playing 8 channels
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/SammyIAm/Moppy/issues/65#issuecomment-42061815
> .

## 

Adam

**Comment by n1cod3mus** [comment]
I have to use logic and coding all day, I'm not sure my brain can handle figuring it out lol but I might have a look over the weekend.

surely though, the pooling cant be set to copy another drive, it would only do the tones if they were overlapping, not the exact same tone?

**Comment by n1cod3mus** [comment]
my arduinos cost me about £7.50 including delivery and each floppy drive was about £5

with cables and connectors you would be looking at at least £3400 to build a 500 floppy machine

**Comment by n1cod3mus** [comment]
oooooorrrr you get everyone who has ever made one, and get them to meet all in the same place, and link them up

**Comment by MultiDaxio** [comment]
Still... floppy fun is so frickin expensive...

2014-05-02 20:28 GMT+02:00 n1cod3mus notifications@github.com:

> my arduinos cost me about £7.50 including delivery and each floppy drive
> was about £5
> 
> with cables and connectors you would be looking at at least £3400
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/SammyIAm/Moppy/issues/65#issuecomment-42062701
> .

## 

Adam

**Comment by MultiDaxio** [comment]
Wait, like Floppy Music Fan Convent/Meeting? Cool!

2014-05-02 20:29 GMT+02:00 Adam Kwiatkowski adikk99@gmail.com:

> Still... floppy fun is so frickin expensive...
> 
> 2014-05-02 20:28 GMT+02:00 n1cod3mus notifications@github.com:
> 
> my arduinos cost me about £7.50 including delivery and each floppy drive
> 
> > was about £5
> > 
> > with cables and connectors you would be looking at at least £3400
> > 
> > —
> > Reply to this email directly or view it on GitHubhttps://github.com/SammyIAm/Moppy/issues/65#issuecomment-42062701
> > .
> 
> ## 
> 
> Adam

## 

Adam

**Comment by n1cod3mus** [comment]
I think i could probably get all of the drives to run from a single UNO, with some other electronics

**Comment by MultiDaxio** [comment]
But then some of the floppies would play the same notes, still pretty cool
though.

2014-05-02 20:33 GMT+02:00 n1cod3mus notifications@github.com:

> I think i could probably get all of the drives to run from a single UNO,
> with some other electronics
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/SammyIAm/Moppy/issues/65#issuecomment-42063209
> .

## 

Adam

**Comment by samphonic** [comment]
You can definitely take advantage of pooling to send multiple notes:

The `DrivePooler.send()` method is actually responsible for passing the (modified) message on to the MIDI Receiver.  But there's no reason it couldn't send several messages to the receiver instead.  The current mapping methods (e.g. `roundRobinMap()`) only return a single message, but you could write one that returns an array of messages, and then loop through them and pass them all to the receiver in the `send()` method.  (I love when abstraction makes things like this easy).

(I got all my drives for free from friends that didn't need floppy drives anymore =P  ).  If you're looking to connect a whole bunch of drives together, keep in mind that with 256 drives, you could produce 8-bit sounds.  But sadly the drives are limited to only about a 400Hz sample rate.  And even at 8-bits, 400Hz sounds are pretty lame.  You might be able to do some cool stuff though.

**Comment by MultiDaxio** [comment]
Work on it more, make floppies your main sound output (who needs
speakes?!), play Mario the way it should be played, ON FLOPPIES!!!!

2014-05-02 20:35 GMT+02:00 SammyIAm notifications@github.com:

> You can definitely take advantage of pooling to send multiple notes:
> 
> The DrivePooler.send() method is actually responsible for passing the
> (modified) message on to the MIDI Receiver. But there's no reason it
> couldn't send several messages to the receiver instead. The current mapping
> methods (e.g. roundRobinMap()) only return a single message, but you
> could write one that returns an array of messages, and then loop through
> them and pass them all to the receiver in the send() method. (I love when
> abstraction makes things like this easy).
> 
> (I got all my drives for free from friends that didn't need floppy drives
> anymore =P ). If you're looking to connect a whole bunch of drives
> together, keep in mind that with 256 drives, you could produce 8-bit
> sounds. But sadly the drives are limited to only about a 400Hz sample rate.
> And even at 8-bits, 400Hz sounds are pretty lame. You might be able to do
> some cool stuff though.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/SammyIAm/Moppy/issues/65#issuecomment-42063498
> .

## 

Adam

**Comment by n1cod3mus** [comment]
I am after a decent copy of the mario theme, all the ones I have come across are all 3 channel for some reason. 

I would like to build a Moppy with just 5.25 drives but sadly on ebay these drive are £35 (GBP) each plus postage so unless i find them somewhere cheap thats not going to happen

**Comment by MultiDaxio** [comment]
Good luck bro! May the Force be with You! :)

2014-05-02 20:50 GMT+02:00 n1cod3mus notifications@github.com:

> I am after a decent copy of the mario theme, all the ones I have come
> across are all 3 channel for some reason.
> 
> I would like to build a Moppy with just 5.25 drives but sadly on ebay
> these drive are £35 (GBP) each plus postage so unless i find them somewhere
> cheap thats not going to happen
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/SammyIAm/Moppy/issues/65#issuecomment-42065047
> .

## 

Adam

**Comment by jazzbassNick** [comment]
@n1cod3mus If you want to attach the midi file you want to use, I'll take a look at it - I may be able to split it up or at least double parts. How many drives you wanna use?

**Comment by MultiDaxio** [comment]
Do not forget about that floppies have slim range of playable notes :)
2 maj 2014 22:30 "jazzbassNick" notifications@github.com napisał(a):

**Comment by n1cod3mus** [comment]
thanks Jazz, I am using Speedy Midi that MultiDaxio put me on to, I have managed to double channels with it, still need to work a little more with it to see what I can do with it. All part of the fun eh.

and just so you know, I'm using 16 floppy drives.

**Comment by jazzbassNick** [comment]
No worries - I'm thinking at least one of the tracks has multiple notes on it for chords, so that splitting that up would get you more channel usage.

I ended up creating a net-new midi file of Funeral March of A Marionette. Got it running, so now I'm planning altering it for more of a visual effect and to better distribute amongst the drives (I'm only running 8).

I've also got an idea of taking the mics out of a bunch of old cordless headsets to set up a mic array so I can individually mix each floppy.

**Comment by n1cod3mus** [comment]
problem you will have is the mics will pick up the drives around them as well

**Comment by MultiDaxio** [comment]
You are on a good way to go! Almost there, good luck!

**Comment by samphonic** [comment]
Closing because there are no lingering "issues".  Free free to continue to use this area for discussion though.
